dvc/repo/brancher.py

```diff
@@ -33,7 +33,10 @@ def brancher(  # noqa: E302
     yield "working tree"
 
     if all_commits:
-        revs = scm.list_all_commits()
+        revs = scm.list_all_commits(
+            exclude_all_tags=not all_tags,
+            exclude_all_branches=not all_branches,
+        )
```

Comment:
This makes `.brancher` and `.used_cache()` interfaces lying ones, `all_commits` no longer means all commits.

Reply:
This also makes `dvc gc --remove-all-branches --remove-all-tags` approaching noop for many scenarios, but not quite noop, a confusing situation. Suppose you make things iteratively and last commits only about some polish like README adjustments.

Reply:
Yeah, the naming `brancher` is at fault here. It should really have been something like `iter_commits()` instead.

But `--remove-all-branches --remove-all-tags` is not a noop, it will exclude tags and branches from the refs for which it iterates through.  Hence why I've mentioned an issue with `refs/remotes` and `refs/stash` etc. We should either exclude those by default or add an option to get rid of those. For now this is a "safe" approach, that can be revisited iteratively.

Reply:
So, it's even more confusing. You need a whole paragraph of text to explain how it works, but still in the many scenarios it's just noop.

Reply:
@Suor Well, that is because of `refs/remotes` and `refs/stash`. Will need to handle them somehow. To me seems like a good idea to just drop them by default and only keep refs/heads and refs/tags. What do you think?